python/datastream/datastream.py

```diff
@@ -0,0 +1,330 @@
+
+    def add_log_info(self, log_dict):
+        for k in log_dict:
+            setattr(self, k, log_dict[k])
+
+    def to_h5(self, directory, filename, log_info):
+        # write datastreams to a file 
+        # filename must include the directory 
+        full_file = os.path.join(directory, filename)
+        with h5py.File(full_file, 'w') as f:
+            grp = f.create_group('metadata') # TODO
+            for dk in self:
+                d_stream = copy.deepcopy(self[dk])
+                dataset = f.create_dataset(d_stream.net, data=d_stream.data) # net must be a string 
+                dstream_attrs = d_stream.__dict__
+                dstream_attrs.pop('data', None)
+                # get attributes from the datastream and set these attributes to the dataset 
+                dataset.attrs.update(dstream_attrs)
+                dataset.attrs.update(log_info) # should log_info itself be a dictionary key?
+            
+            grp.attrs['json'] = json.dumps(self.__dict__, default=vars)
+
+
+def h5_to_datastreams(directory, filename):
+
+    full_file = os.path.join(directory, filename)
+
+    datastreams = Datastreams()
+    with h5py.File(full_file, "r") as file:
+        for dk in file.keys():
+            if dk == 'metadata':
+                t2 = json.loads(file[dk].attrs['json'])
+                datastreams.add_log_info(t2)
+            else:
+                try:
+                    ds = Datastream(file[dk][:],  # must use [:] to create a copy of the data so that the h5 file doesn't need to stay open
+                                sample_rate=file[dk].attrs['sample_rate'], 
+                                units=file[dk].attrs['units'], 
+                                name=file[dk].attrs['name'], 
+                                net=file[dk].attrs['net'], 
+                                t0=file[dk].attrs['initial_time'])
+
+                    datastreams[ds.net] = ds
+                except:
+                    print(f'skipping h5 key {dk}')
+
+    return datastreams
+
+
+
+def rawh5_to_datastreams(data_dir, infile, data_to_names, daq, phys_connections, outfile = None):
+    """
+    parameters: 
+        data_dir: directory for both input and output files 
+        infile: name of input file (needs .h5 suffix)
+        data_to_names: this is a method of the DDR, pass this in to complete
+        daq: the DAQ board object (to get info such as sequencer_setup)
+        phys_connection: dictionary of PhysicalConnections 
+        outfile: name of output h5 file 
+
+    returns: 
+        datastreams 
+
+    """
+
+    # to get the deswizzled data of all repeats need to read the file
+    _, chan_data = read_h5(data_dir, file_name=infile, chan_list=np.arange(8))
+
+    # Long data sequence -- entire file
+    adc_data, timestamp, dac_data, ads_data_tmp, ads_seq_cnt, read_errors = data_to_names(chan_data)
+    ads_separate_data = extract_ads_data(ads_data_tmp, ads_seq_cnt, daq.ADC_gp.sequencer_setup)
```

Comment:
This is throwing an error for me. Do we need to add this to [pyripherals](https://github.com/Ajstros/pyripherals/blob/main/python/src/pyripherals/peripherals/ADS8686.py#L137) first?
```term
Traceback (most recent call last):
  File "bathclamp_vclamp_step_response.py", line 382, in <module>
    datastreams, log_info = rawh5_to_datastreams(data_dir, infile, ddr.data_to_names, daq, sys_connections, outfile = None)
  File "datastream.py", line 160, in rawh5_to_datastreams
    ads_separate_data = extract_ads_data(ads_data_tmp, ads_seq_cnt, daq.ADC_gp.sequencer_setup)
AttributeError: 'ADS8686' object has no attribute 'sequencer_setup'
```